src/helpers/formulaHelper.js

```diff
@@ -0,0 +1,187 @@
+const formulas = require('./formulas.json')
+
+const TOKEN_TYPE = {
+    Number: Symbol("number"),
+    Variable: Symbol("variable"),
+    Function: Symbol("function"),
+    Operator: Symbol("operator")
+};
+
+const FUNCTION_TOKENS = [
+    "MIN",
+    "MAX",
+    "ROUND"
+];
+
+function getFormulaByKey(key) {
+    return formulas[key];
+}
+
+function parse(formulaObj) {
+    const constants = formulaObj.constants;
+    let tokens = formulaObj.formula.split('');
+
+    let resultStack = [];
+    let parseStack = [];
+
+    for (let i = 0; i < tokens.length; i++) {
+        if (tokens[i] == ' ')
+            continue;
+        
+        if (tokens[i] >= '0' && tokens[i] <= '9') {
+            let number = "";
+            while (i < tokens.length && tokens[i] >= '0' && tokens[i] <= '9') {
+                number += tokens[i];
+                i++;
+            }
+            resultStack.push({ token: parseInt(number), type: TOKEN_TYPE.Number });
+            i--;
+        }
+
+        else if (isValidNonNumberToken(tokens[i])) {
+            let tempToken = "";
+            while (i < tokens.length &&
+                (isValidNonNumberToken(tokens[i]) || (tokens[i] >= '0' && tokens[i] <= '9'))) 
+            {
+                tempToken += tokens[i];
+                i++;
+            }
+            if (constants.hasOwnProperty(tempToken)) {
+                resultStack.push({ token: constants[tempToken], type: TOKEN_TYPE.Number });
+            } else if (FUNCTION_TOKENS.includes(tempToken.toUpperCase())) {
+                parseStack.push(tempToken.toUpperCase());
+            } else {
+                resultStack.push({ token: tempToken, type: TOKEN_TYPE.Variable });
+            }
+
+            i--;
+        }
+
+        else if (tokens[i] == '(' || tokens[i] == ',') {
+            parseStack.push(tokens[i]);
+        }
+
+        else if (tokens[i] == ')') {
+            for (let char = parseStack.pop(); char != '('; char = parseStack.pop()) {
+                pushOperator(resultStack, char);
+            }
+            if (FUNCTION_TOKENS.includes(parseStack[parseStack.length - 1])) {
+                pushOperator(resultStack, parseStack.pop());
+            }
+        }
+
+        else {
+            while (parseStack.length > 0 &&
+                precedence(tokens[i]) <= precedence(parseStack[parseStack.length - 1])
+            ) {
+                pushOperator(resultStack, parseStack.pop());
+            }
+            parseStack.push(tokens[i]);
+        }
+    }
+
+    while (parseStack.length > 0) {
+        let operator = parseStack.pop();
+        pushOperator(resultStack, operator);
+    }
+
+    return resultStack;
+}
+
+function isValidNonNumberToken(char) {
+    return (
+        (char >= 'a' && char <= 'z') || 
+        (char >= 'A' && char <= 'Z') ||
+        (char == '_')
```

Comment:
Constants galore~

Reply:
Hmm I don't think this one needs to be extracted since the function already explains what it does. Same case to comparing character between `0` and `9`.

Reply:
Fair enough, resolving